Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,357 @@
+                group.cancelAll()
+                return result
+            } catch {
+                group.cancelAll()
+                throw error
+            }
+        }
+    }
+
+    /// Runs the event loop and moves "time" forward to the given point in time, running any scheduled
+    /// tasks that need to be run.
+    ///
+    /// - Note: If `deadline` is before the current time, the current time will not be advanced.
+    public func advanceTime(to deadline: NIODeadline) async {
+        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
+            self.queue.async {
+                let newTime = max(deadline, self.now)
+
+                while let nextTask = self.scheduledTasks.peek() {
+                    guard nextTask.readyTime <= newTime else {
+                        break
+                    }
+
+                    // Now we want to grab all tasks that are ready to execute at the same
+                    // time as the first.
+                    var tasks = Array<EmbeddedScheduledTask>()
+                    while let candidateTask = self.scheduledTasks.peek(), candidateTask.readyTime == nextTask.readyTime {
+                        tasks.append(candidateTask)
+                        self.scheduledTasks.pop()
+                    }
+
+                    // Set the time correctly before we call into user code, then
+                    // call in for all tasks.
+                    self._now.store(nextTask.readyTime.uptimeNanoseconds)
+
+                    for task in tasks {
+                        task.task()
+                    }
+                }
+
+                // Finally ensure we got the time right.
+                self._now.store(newTime.uptimeNanoseconds)
+
+                continuation.resume()
+            }
+        }
+    }
+
+    /// Executes the given function in the context of this actor. This is useful when it's necessary to be confident that an operation
+    /// is "blocking" the actor. As long as you are executing, nothing else can execute in this loop.
+    public func executeInContext<ReturnType: Sendable>(_ task: @escaping @Sendable () throws -> ReturnType) async throws -> ReturnType {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<ReturnType, Error>) in
+            self.queue.async {
+                do {
+                    continuation.resume(returning: try task())
+                } catch {
+                    continuation.resume(throwing: error)
+                }
+            }
+        }
+    }
+
+    internal func drainScheduledTasksByRunningAllCurrentlyScheduledTasks() {
+        var currentlyScheduledTasks = self.scheduledTasks
+        while let nextTask = currentlyScheduledTasks.pop() {
+            self._now.store(nextTask.readyTime.uptimeNanoseconds)
+            nextTask.task()
+        }
+        // Just fail all the remaining scheduled tasks. Despite having run all the tasks that were
```

Comment:
IIRC MTELG allows some number of tasks to be executed if they were scheduled as a result of another task running. Are we okay diverging from that behaviour?

Reply:
For now, I think yes. This is consistent with `EmbeddedEventLoop`.